core/src/messages/agent.rs

```diff
@@ -149,53 +157,57 @@ impl BackendStatsMessage {
         cur_stats_message: &Stats,
     ) -> Result<BackendStatsMessage, Error> {
         // Based on docs here: https://docs.docker.com/engine/api/v1.41/#tag/Container/operation/ContainerStats
-
-        use anyhow::Context;
-        let mem_naive_usage = cur_stats_message
-            .memory_stats
-            .usage
-            .context("no memory stats.usage")?;
-        let mem_available = cur_stats_message
-            .memory_stats
-            .limit
-            .context("no memory stats.limit")?;
-        let mem_stats = cur_stats_message
-            .memory_stats
-            .stats
-            .context("no memory stats.stats")?;
+        let mem_naive_usage = cur_stats_message.memory_stats.usage;
+        let mem_available = cur_stats_message.memory_stats.limit;
+        let mem_stats = cur_stats_message.memory_stats.stats;
         let cache_mem = match mem_stats {
-            bollard::container::MemoryStatsStats::V1(stats) => stats.cache,
-            bollard::container::MemoryStatsStats::V2(stats) => stats.inactive_file,
+            Some(MemoryStatsStats::V1(stats)) => Some(stats.cache),
+            Some(MemoryStatsStats::V2(stats)) => Some(stats.inactive_file),
+            _ => None,
+        };
+        let mem_used = match (mem_naive_usage, cache_mem) {
+            (Some(mem_naive_usage), Some(cache_mem)) => Some(mem_naive_usage - cache_mem),
+            _ => None,
+        };
+        let mem_utilization = match (mem_used, mem_available) {
+            (Some(mem_used), Some(mem_available)) => Some(mem_used as f64 / mem_available as f64),
+            _ => None,
```

Comment:
I felt it was better to send a stats message of some sort here rather than error out if docker stats didn't return a value for any one of these. I could be persuaded to switch back, though, as I'm not entirely sure in what scenarios these values could come back as `None`s. 🤔